### ISSUE TITLE
Some (rare) shapes are visibly distorted

### ISSUE THREAD

**Issue by Self-Perfection** [osmandapp/OsmAnd#9821]
# 🐞 bug report

### Is this a regression?
Nope. This building was distorted in OsmAnd as long as I recall, more than 6 months.

### Description
Some shapes are drastically distorted in OsmAnd maps in comparison to corresponding OSM object. Here is the most notorious example I've seen:
![Screenshot_20200916-121116](https://user-images.githubusercontent.com/3757038/93326880-2fa2db00-f822-11ea-8334-f64d0899b149.png)
Compare shapes of buildings [130 к3](https://www.openstreetmap.org/way/189195301) (upper left corner) and [128 к2](https://www.openstreetmap.org/way/188114318) (lower right). Both are rather accurate circles in OSM and yet OsmAnd shows 130 к3 as polygon with just a few nodes and sharp and distinct angles.

## 🔬 Minimal Reproduction
Download Russia/Moscow map, zoom to [example building](https://www.openstreetmap.org/way/189195301).

## 🔥 Exception or Error

Not applicable

## 🌍 Your Environment

**OsmAnd Version:**

<pre><code>
OsmAnd 3.7.4

</code></pre>

**Device and Android/iOS version:**
Xiaomi Mi A1
Android 9
**Maps used (online or offline):**

- [x] Offline maps offered within the OsmAnd app for download.

- [ ] Online (tile / raster) maps 

**Anything else relevant?**

**Comment by Famlam** [comment]
It seems this especially occurs with circles. Some other examples:
- old tower @ 51.84747°N 5.85759°E
- grass @ 51.84872°N 5.86250°E

**Comment by vshcherb** [comment]
Merged to https://github.com/osmandapp/OsmAnd-Issues/issues/1049

**Comment by fntagit** [comment]
Distorsions still happen on version 4.2.0.
![Capture d’écran   2022-05-19 à 09 48 55](https://user-images.githubusercontent.com/42489715/169558411-bfb3faa0-98b6-4742-928c-c63606d316ca.png)

**Comment by fntagit** [comment]
Distorsions are not rare at all, and affect all kind of shapes. It seems that the nodes are not placed exactly, and thus the shape of the ways is distorted. Here is another example, where all shapes are distorted, with a screen shot from Go Map!! showing the correct shapes.
![E19BB72E-3214-4C09-8DC6-E33811AA5481](https://user-images.githubusercontent.com/42489715/169650443-af474593-70e8-473f-8ff7-5cb03f4eddce.png)
![C6813C24-1681-4CAC-BEEE-C4135E1198F1](https://user-images.githubusercontent.com/42489715/169650447-ccd24a97-037d-4a84-9ba8-daeb191a0a53.png)

**Comment by Famlam** [comment]
@fntagit I think your cases might be due to the coordinates of i.e. nodes being rounded (e.g. only affecting high zoom levels), not sure if it's the same distortion as the first post

**Comment by fntagit** [comment]
Indeed, rounding, or maybe truncating, errors are a very likely cause. Now, the question is: where are those errors introduced? Is it in the App, when computing the transformation between global and screen coordinates, or in the OsmAnd servers, when processing the OSM data to produce the dowloadable maps.

**Comment by vshcherb** [comment]
@fntagit they are not rare or frequent, distortions happen up to 1 m precision, so in case mappers use micromapping everywhere, the distortions will be everywhere

**Comment by dreirund** [comment]
> *distortions happen up to 1 m precision, so in case mappers use micromapping everywhere, the distortions will be everywhere*

_Should_ they happen or not? Why do they happen, is it feasible to fix it?

**Comment by Famlam** [comment]
Maybe leave this issue for the large distortions and open a separate issue for the micromapping distortion?

**Comment by vshcherb** [comment]
All distortions are now the same. We align to zoom 26th zoom which corresponds pixel on 18th zoom (https://wiki.openstreetmap.org/wiki/Zoom_levels ) - 0.596m. Cause map shifts are hardcoded it would require massive map data refactoring, new map generations and will be problematic for already existing installations. 

So it's unlikely to happen in near future, we can talk to make some specific objects more precise but we need to categorize these objects somehow.

**Comment by ronnystandtke** [comment]
With more and more tiny and high-precision positioned areas added to the OSM maps (e.g. [single parking spaces](https://wiki.openstreetmap.org/wiki/Tag:amenity%3Dparking_space) or highly detailed building data provided by the government), this issue becomes visible all over the place. The map looks really great on the OSM web page but horrible in the OsmAnd app.
The internal grid of only 1 meter precision was perhaps a good idea at the beginning of OsmAnd, but should certainly be refined now. Maybe with a transition period where the old low precision maps and new high precision maps are provided in parallel?

**Comment by dreirund** [comment]
But can you at least fix it in the sense that topologically objects are correct?

i.e. even if you deform the objects, make sure that touching edges still touch, non-overlapping stuff does not overlapp, and touching lines still touch?

When you aligh nodes on a 1&thinsp;m grid, you could still guarantee topoligical correctness.

But [the topological correctness is not fulfilled, too](https://github.com/osmandapp/OsmAnd/issues/15317).
